### PR TITLE
[Testing] Fix notifier flakey tests

### PR DESCRIPTION
**Pull request by zhangchiqing**
This PR fixes a race condition in notifier tests, that we started all the workers first before the consumers are ready, this could cause an edge case which causes flakey tests that consumers missed all notifications because they haven't started. This explains why the flakey tests sometimes failed with `only consumed 0 units of work but expecting 100`.

The fix is to start the consumers first, wait until they are all ready (listening to channels). Then start all workers, wait until they all started. Such that, the edge case won't happen again.

### CODE REVIEW

File: engine/notifier_test.go [hunk 1]
```diff
@@ -159,8 +151,25 @@ func TestNotifier_AllWorkProcessed(t *testing.T) {
...
+				workersAllReady.Done()
+				for scheduledWork.Inc() <= totalWork {
+					pendingWorkQueue <- struct{}{}
+					notifier.Notify()
+				}
+			}()
+		}
+
+		// wait long enough for all workers to be started.
+		consumersAllReady.Wait()
```
Comment (smnzhu) [inline engine/notifier_test.go:172]: Should be workersAllReady.Wait

Comment (zhangchiqing) [reply engine/notifier_test.go]: Good catch!

**Comment by smnzhu** [comment]
Thanks Leo! Can you add back the `t.Parallel` call, and try running with 500 iterations as described [here](https://github.com/onflow/flow-go/pull/1037#issuecomment-888017485) to make sure it passes? Thanks :)

**Comment by zhangchiqing** [comment]
[t.Parallel](https://github.com/onflow/flow-go/blob/66cf6359a6a63e6e0b3695c52624f0b242b19993/engine/notifier_test.go#L123) is there, I didn't remove.

I updated to run 500 locally, and it passed. I'm not sure we want to let CI run 500 times. 
I think 100 is enough, with 10 PRs, it will be ran for 1000 times.

File: engine/notifier_test.go [hunk 2]
```diff
@@ -107,11 +107,15 @@ func TestNotifier_ManyConsumers(t *testing.T) {
...
 	// require that all workers got a notification
 	require.Eventuallyf(t,
-		func() bool { return pendingWorkers.Load() == 0 },
+		func() bool {
+			nPendingWorkers = pendingWorkers.Load()
+			return nPendingWorkers == 0
+		},
 		3*time.Second, 100*time.Millisecond,
-		"still awaiting %d workers to get notification", pendingWorkers.Load(),
+		"still awaiting %d workers to get notification", nPendingWorkers,
```
Comment (smnzhu) [inline engine/notifier_test.go:118]: Perhaps you still misunderstand my point about the error comments from earlier? This is exactly the opposite of what we want :/

File: engine/notifier_test.go [hunk 3]
```diff
@@ -159,14 +155,35 @@ func TestNotifier_AllWorkProcessed(t *testing.T) {
...
 		// require that all work is eventually consumed
 		require.Eventuallyf(t,
-			func() bool { return consumedWork.Load() == totalWork },
+			func() bool {
+				nConsumedWork = consumedWork.Load()
+				return nConsumedWork == totalWork
+			},
 			3*time.Second, 100*time.Millisecond,
-			"only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork,
+			"only consumed %d units of work but expecting %d", nConsumedWork, totalWork,
```
Comment (smnzhu) [inline engine/notifier_test.go:186]: Same here

File: engine/notifier_test.go (hunk 2 above)
Comment (zhangchiqing) [reply engine/notifier_test.go:118]: Note, at Line 114, we update the pendingWorks every time we check the condition, if the condition in the end still didn't meet, then the cached `nPendingWorks` is the most recent, and we will read `nPendingWorks` at that moment.

require.Eventually takes in the variable not the value, it won't read the variable until waited for timeout.

**Comment by smnzhu** [comment]
I've cherry picked the relevant commits into https://github.com/onflow/flow-go/pull/1037, so I'll close this one for now :)

**Pull request closed without merging**